File: 1683-maximum-number-of-coins-you-can-get/maximum-number-of-coins-you-can-get.py
class Solution(object):
    def maxCoins(self, piles):
        max_val = max(piles)
        freq = [0] * (max_val + 1)

        for i in piles:
            freq[i] += 1

        coins = 0
        chance = len(piles) // 3
        turn = 1
        i = max_val

        while chance != 0:
            if freq[i] > 0:
                if turn == 1:
                    turn = 0
                else:
                    chance -= 1
                    turn = 1
                    coins += i
                freq[i] -= 1
            else:
                i -= 1

        return coins

File: 1683-maximum-number-of-coins-you-can-get/maximum-number-of-coins-you-can-get.js
var maxCoins = function(piles) {
    let max = 0;
    let n = piles.length;

    for (let i of piles) {
        if (max < i) max = i;
    }

    let freq = new Array(max + 1).fill(0);

    for (let i of piles) {
        freq[i]++;
    }

    let coins = 0;
    let chance = Math.floor(n / 3);
    let turn = 1;
    let i = max;

    while (chance !== 0) {
        if (freq[i] > 0) {
            if (turn === 1) turn = 0;
            else {
                chance--;
                turn = 1;
                coins += i;
            }
            freq[i]--;
        } else {
            i--;
        }
    }

    return coins;
};